common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +173,445 @@
+            try {
+                S newInstance = initInstance(clazz, argTypes, args);
+                return newInstance;
+            } catch (Throwable t) {
+                throw new IllegalStateException("Extension instance(definition: " + definition + ", class: " +
+                        type + ")  could not be instantiated: " + t.getMessage(), t);
+            }
         }
 
-        if (extensions.isEmpty()) {
-            return extensions;
-        }
-        extensions.sort((c1, c2) -> {
-            int o1 = 0;
-            int o2 = 0;
-            @SuppressWarnings("unchecked")
-            LoadLevel a1 = (LoadLevel) c1.getAnnotation(LoadLevel.class);
-            @SuppressWarnings("unchecked")
-            LoadLevel a2 = (LoadLevel) c2.getAnnotation(LoadLevel.class);
-
-            if (a1 != null) {
-                o1 = a1.order();
+        private List<Class> loadAllExtensionClass(ClassLoader loader) {
+            List<Class> result;
+            List<ExtensionDefinition> definitions = definitionsHolder.get();
+            if (definitions == null) {
+                synchronized (definitionsHolder) {
+                    definitions = definitionsHolder.get();
+                    if (definitions == null) {
+                        definitions = findAllExtensionDefinition(loader);
+                        definitionsHolder.set(definitions);
+                    }
+                }
             }
+            result = definitions.stream().map(def -> def.getServiceClass()).collect(Collectors.toList());
+            return result;
+        }
 
-            if (a2 != null) {
-                o2 = a2.order();
+        @SuppressWarnings("rawtypes")
+        private List<ExtensionDefinition> findAllExtensionDefinition(ClassLoader loader) {
+            List<ExtensionDefinition> extensionDefinitions = new ArrayList<>();
+            try {
+                loadFile(SERVICES_DIRECTORY, loader, extensionDefinitions);
+                loadFile(SEATA_DIRECTORY, loader, extensionDefinitions);
+            } catch (IOException e) {
+                throw new EnhancedServiceNotFoundException(e);
             }
 
-            return Integer.compare(o1, o2);
-
-        });
+            //After loaded all the extensions,sort the caches by order
+            if (!nameToDefinitionsMap.isEmpty()) {
+                for (List<ExtensionDefinition> definitions : nameToDefinitionsMap.values()) {
+                    Collections.sort(definitions, (def1, def2) -> {
+                        int o1 = def1.getOrder();
+                        int o2 = def2.getOrder();
+                        return Integer.compare(o1, o2);
+                    });
+                }
+            }
 
-        return extensions;
-    }
+            if (!extensionDefinitions.isEmpty()) {
+                Collections.sort(extensionDefinitions, (definition1, definition2) -> {
+                    int o1 = definition1.getOrder();
```

Comment:
Since extensionDefinitions were sorted, is the defaultExtensionDefinition calculating in `getExtensionDefinition` 
 necessary?

Reply:
Yeah, it is true that the calculating in getExtensionDef is  unnecessary,I will remove them.